### PR TITLE
Winch: Implement rem for aarch64

### PR DESCRIPTION
**Pull request by MarinPostma**
Implement remainder operation for `aarch64` in winch.

one more step in the #8321 series

### CODE REVIEW

File: winch/codegen/src/isa/aarch64/asm.rs [hunk 1]
```diff
@@ -473,6 +475,67 @@ impl Assembler {
...
+        self.emit(Inst::TrapIf {
+            kind: CondBrKind::Zero(divisor.into()),
+            trap_code: TrapCode::INTEGER_DIVISION_BY_ZERO,
+        });
+
+        // `cranelift-codegen` doesn't support emitting u/sdiv for anything but I64,
+        // we therefore sign-extend the operand.
+        // see: https://github.com/bytecodealliance/wasmtime/issues/9766
+        //
+        // FIXME: maybe this is only necessary for signed remainder?
```
Comment (MarinPostma) [inline winch/codegen/src/isa/aarch64/asm.rs:497]: @saulecabrera I'm not sure about that, but if we are performing an unsigned remainder operation, then do we really care about sign extending? Maybe it's not necessary for division either? Is there a scenario where we can't just blindly consider the `w` register as a `x` register?

Comment (MarinPostma) [reply winch/codegen/src/isa/aarch64/asm.rs:497]: The same reasonning applies to div, I think. We could shave a few instructions in the case of unsigned division too.

File: winch/codegen/src/isa/aarch64/asm.rs [hunk 2]
```diff
@@ -473,6 +475,67 @@ impl Assembler {
...
+    pub fn rem_rrr(
+        &mut self,
+        divisor: Reg,
+        dividend: Reg,
+        dest: Writable<Reg>,
+        kind: RemKind,
+        size: OperandSize,
+    ) {
+        // Check for division by 0
+        self.emit(Inst::TrapIf {
```
[the remark this answers is not in the document]
Comment (saulecabrera) [reply winch/codegen/src/isa/aarch64/asm.rs:488]: Also, the same invariants as signed division apply here, right? i.e., we must still check for overflow.

**Review by saulecabrera — COMMENTED** [review]
Looking good. Left some comments inline.

As a general comment, given that the same invariant as division apply here, could we modify the existing `div_rrr` so that it can be re-used from `rem_rrr`?

File: winch/codegen/src/isa/aarch64/asm.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (MarinPostma) [reply winch/codegen/src/isa/aarch64/asm.rs:488]: Actually I was expecting it to be the case, but as it turn out, it doesn't seem necessary; the following snippet
```wat
 (module
     (func (param i32 i32) (result i32)
          local.get 0
           local.get 1
           i32.rem_u
     )
 )
```

is compiled by cranelift to:

```armasm
0000000000000000 <wasm[0]::function[0]>:
       0: d503235f      pacibz
       4: a9bf7bfd      stp     x29, x30, [sp, #-16]!
       8: 910003fd      mov     x29, sp
       c: 2a0403e3      mov     w3, w4
      10: 2a0503e5      mov     w5, w5
      14: b40000c5      cbz     x5, 0x2c <wasm[0]::function[0]+0x2c>
      18: 9ac50868      udiv    x8, x3, x5
      1c: 9b058d02      msub    x2, x8, x5, x3
      20: a8c17bfd      ldp     x29, x30, [sp], #16
      24: d50323df      autibz
      28: d65f03c0      ret
      2c: 0000c11f      udf     #49439
```

(with `rem_s`:
```armasm
       0: d503235f      pacibz
       4: a9bf7bfd      stp     x29, x30, [sp, #-16]!
       8: 910003fd      mov     x29, sp
       c: 52b00004      mov     w4, #-2147483648
      10: 93407c81      sxtw    x1, w4
      14: 92800003      mov     x3, #-1
      18: 9ac30c25      sdiv    x5, x1, x3
      1c: 9b0384a2      msub    x2, x5, x3, x1
      20: a8c17bfd      ldp     x29, x30, [sp], #16
      24: d50323df      autibz
      28: d65f03c0      ret
```
)

At first I thought this was a bug, but it seems to be consistent across platform:
```wat
(module
    (func (export "_start") (result i32)
	(i32.const 0x80000000)
	(i32.const -1)
        i32.rem_s
    )
)
```

this yields `0` on both aarch64 and x86.

**Comment by MarinPostma** [comment]
I was my initial intention to factorize the check, but as it turns out, we don't need to check for overflow  for `rem`. I'm working on supporting i32 division, so we can drop the  extension in a subsequent PR too.

File: winch/codegen/src/isa/aarch64/asm.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (saulecabrera) [reply winch/codegen/src/isa/aarch64/asm.rs]: Ah right, going over the Wasm spec, there's no explicit mention of overflow checks in this case.

**Pull request merged**